ion-hash/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+
+//! Implement Ion Hash as per the [spec](https://amzn.github.io/ion-hash/docs/spec.html)
+//!
+//! ## Examples
+//! ```rust
+//! use ion_rs::value::loader::{self, Loader};
+//! use ion_rs::result::IonResult;
+//!
+//! # fn main() -> IonResult<()> {
+//!   let loader = loader::loader();
+//!   let elem = loader.iterate_over(b"\"hello world\"")?.next().unwrap()?;
+//!   let digest = ion_hash::sha256(&elem);
+//!   println!("{:?}", digest);
+//! # Ok(())
+//! # }
+//! ```
+
+use ion_rs::{
+    value::{owned::OwnedElement, Element},
+    IonType,
+};
+
+use sha2::{digest::Output, Digest, Sha256};
+
+pub fn sha256(elem: &OwnedElement) -> Vec<u8> {
+    let mut hasher = IonHasher::new(Sha256::new());
+    let result = hasher.visit_element(elem);
+    Vec::from(result.as_slice())
+}
+
+/// Bytes markers as per the spec.
+mod markers {
+    /// single byte begin marker
+    pub(crate) const B: u8 = 0x0B;
+    /// single byte end marker
+    pub(crate) const E: u8 = 0x0E;
+    /// single byte escape
+    pub(crate) const ESC: u8 = 0x0C;
+    /// type qualifier octet consisting of a four-bit type code T followed by a four-bit qualifier Q
+    /// (this varies per type)
+    pub(crate) const TQ_STRING: u8 = 0x80;
+}
+
+// TODO: In the fullness of time, we will have a IonHashReader and a
+// IonHashWriter. This will allow reading/writing a value *while* computing a
+// hash. For now, we provide a standalone hasher using the OwnedElement API.
+pub struct IonHasher<D>
+where
+    D: Digest,
+{
+    hasher: D,
+}
+
+impl<D> IonHasher<D>
+where
+    D: Digest,
+{
+    fn new(hasher: D) -> IonHasher<D> {
+        IonHasher { hasher }
+    }
+
+    fn visit_element(&mut self, elem: &OwnedElement) -> Output<D> {
+        let serialized_bytes = match elem.ion_type() {
```

Comment:
I think this should be made generic and I think this avoids the dependency on `OwnedElement` (which is good).

```rust
fn hash_element<E: Element + ?Sized>(&mut self, elem: &E) -> Output<D>{ ... }
```

Any thoughts about making `D` a parameter and allowing `IonHasher` to be used in a stateless way?

E.g.

```rust
/// Constructs a [`Digest`] and performs a self-contained hash for a given [`Element`].
fn hash_element<D: Digest, E: Element + ?Sized>(&self, elem: &E) -> Output<D> {
    let mut digest = D::new();
    self.update_hash_element(&mut digest, elem);
    digest.finalize()
}

/// Hashes an element and updates an existing [`Digest`].
fn update_hash_element<D: Digest, E: Element + ?Sized>(&self, digest: &mut D, elem: &E) { ... }
```

This would take the type parameter away from the `IonHasher` and push the parameterization on the invocation, but it would also decouple the usage of the digest (especially around finalization, if the user wants to do other things before `finalize`).  That said, `finalize_reset` might be more optimal for subsequent hashing, so that might be an argument to keep it internalized (and that doesn't mean we couldn't add the update method later).

Reply:
> I think this should be made generic and I think this avoids the dependency on OwnedElement (which is good).

Done!

> Any thoughts about making D a parameter and allowing IonHasher to be used in a stateless way?

I'm expecting some churn here in our API; as we use it we'll learn. For now I'm happy to leave as is or take your suggestion - pretty ambivalent.

Reply:
Let's not change it--I was more curious then anything.